### PR TITLE
Add getNextDistributionTime function

### PR DESCRIPTION
**Pull request by shreyaspapi**

### CODE REVIEW

File: contracts/REXUniswapV3Market.sol [hunk 1]
```diff
@@ -1008,6 +1008,23 @@ contract REXUniswapV3Market is
...
+     *
+     * @param gasPrice The gas price in wei per gas unit for the transaction.
+     * @param gasLimit The maximum amount of gas to be used for the transaction.
+     * @param inflowRate The rate of incoming tokens per second.
+     * @param lastDistributedAt The timestamp of the last token distribution.
+     * @param tokenToMaticRate The conversion rate from tokens to Matic.
+     *
+     * @return The timestamp for the next token distribution.
+     */
+    function getNextDistributionTime(uint256 gasPrice, uint256 gasLimit, uint256 inflowRate, uint256 lastDistributedAt, uint256 tokenToMaticRate) public view returns (uint256) {
```
Comment (mikeghen) [inline contracts/REXUniswapV3Market.sol]: Could you look into computing the inflowRate internally with a `cfa` call?

And, could you remove lastDistributedAt as a parameter on this method and instead use the contract variable?

File: test/REXUniswapV3Market.test.ts [hunk 2]
```diff
@@ -1515,5 +1515,25 @@ describe('REXUniswapV3Market', () => {
         })
         .exec(bobSigner)
     })
+
+    it('#4.2 Should return the correct next distribution time', async () => {
+      const gasPrice = 320 // 300 GWEI
+      const gasLimit = 120000
+      const inflowRate = 10555 // infow in gwei -> 0.38 usd / hr
+      const lastDistributedAt = 1651261812
```
Comment (mikeghen) [inline test/REXUniswapV3Market.test.ts]: Same here, I think `inflowRate` and `lastDistributedAt` can be gotten internal to the method vs. as an argument to the function.

**Review by mikeghen — CHANGES_REQUESTED** [review]
There might be a reason not to use the `cfa` to get the `inflowRate` and maybe a good reason to make it an argument, I would have expected the `getNextDistributionTime` method would not accept those too arguments and instead would be gotten in the method, here's a document on getting the netFlowRate for a token at the solidity contract level: https://docs.superfluid.finance/superfluid/developers/constant-flow-agreement-cfa/cfav1-library#getting-stream-data

Let me know what you think @shreyaspapi

**Comment by shreyaspapi** [comment]
@mikeghen Made the requested changes

File: contracts/REXUniswapV3Market.sol [hunk 3]
```diff
@@ -1008,6 +1008,23 @@ contract REXUniswapV3Market is
...
+     *
+     * @return The timestamp for the next token distribution.
+     */
+    function getNextDistributionTime(uint256 gasPrice, uint256 gasLimit, uint256 tokenToMaticRate) public view returns (uint256) {
+        uint256 inflowRate = uint256(int256(cfa.getNetFlow(inputToken, address(this)))) / (10 ** 9); // Safe conversion - Netflow rate will always we positive or zero
+        
+        uint256 tokenAmount = gasPrice * gasLimit * tokenToMaticRate;
+        uint256 timeToDistribute = (tokenAmount / inflowRate) / (10 ** 9);
+        return lastDistributedAt + timeToDistribute;
+    }
```
Comment (mikeghen) [inline contracts/REXUniswapV3Market.sol:1026]: Looks good to me :+1:

File: test/REXUniswapV3Market.test.ts [hunk 4]
```diff
@@ -1515,5 +1515,24 @@ describe('REXUniswapV3Market', () => {
...
+        superToken: ricochetMATICx.address,
+        account: market.address,
+        providerOrSigner: adminSigner,
+      })
+
+      console.log(netFlowRate.toString())
+
+      const actualDistributionTime = await market.getNextDistributionTime(gasPrice, gasLimit, tokenToMaticRate)
+
+      expect(actualDistributionTime).to.greaterThan(lastDistributedAt)
```
Comment (mikeghen) [inline test/REXUniswapV3Market.test.ts]: can this be more precise like:
```
      expect(actualDistributionTime).to.equal(lastDistributedAt + gasPrice * gasLimit * tokenToMaticRate / inflowRate)
```

**Review by mikeghen — CHANGES_REQUESTED** [review]

File: test/REXUniswapV3Market.test.ts (hunk 4 above)
Comment (shreyaspapi) [reply test/REXUniswapV3Market.test.ts]: Done👍, Added `calculatedDistributionTime` instead of using `greaterThan`.

**Review by mikeghen — APPROVED** [review]
Looking good 👏

**Pull request merged**